sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/util/GenericArrayDataHelper.scala

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+object GenericArrayDataHelper {
+  def toArray(seq: Seq[Any]): Array[Any] = seq.toArray
+
+  def toArray(seqOrArray: Any): Array[Any] = seqOrArray match {
+    case seq: Seq[Any] => seq.toArray
+    case array: Array[Any] => array  // array of objects, so no need to convert
+    case array: Array[_] => array.toSeq.toArray[Any] // array of primitives, so box them
```

Comment:
Does this actually box? On Scala 2.13, which I have handy, this just fails when accessing the result. I guess this is how the current code works, so hey

Reply:
To be honest, I didn't go deep into the specific code that it was called by. This class just extracted the original code into this helper class without any changes. The actual changes is in `scala-2.13/org/apache/spark/sql/catalyst/util/GenericArrayDataHelper.scala`.

If need to clarify the answer to this question, I need to investigate it further

Reply:
Yes, I it is not `box`, change to `type conversion` instead? Or do you have any other suggestions?

Reply:
I'm just not sure if it works in 2.12, but maybe it does. It doesn't work in 2.13 but maybe I'm misunderstanding it. Like, it would have never worked in Spark otherwise it seems, so I'm probably wrong.

Reply:
> I'm just not sure if it works in 2.12, but maybe it does. It doesn't work in 2.13 but maybe I'm misunderstanding it. Like, it would have never worked in Spark otherwise it seems, so I'm probably wrong.

OK, I know what you mean. Let me track this and if this is an issue, I will file a new Jira.

Reply:
> I'm just not sure if it works in 2.12, but maybe it does. It doesn't work in 2.13 but maybe I'm misunderstanding it. Like, it would have never worked in Spark otherwise it seems, so I'm probably wrong.

@srowen I think it works well, both in Scala 2.12 and Scala 2.13.

For example, it is used to construct `ArrayBasedMapData`

https://github.com/apache/spark/blob/e4ca8424474e571d8e137388fe5d54732b68c2f3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ArrayBasedMapData.scala#L119-L127

The above apply method is used directly or indirectly in many places in Spark code, I have added logs to verify that many cases have entered the `case array: Array[_] => array.toSeq.toArray[Any]` branch

Reply:
In fact, maybe we should be able to simplify `array.toSeq.toArray[Any]` to `array.toArray[Any]`. There are many other similar cases in `GenericArrayData`. 
This may be just a simplification for Scala 2.12 because `toSeq` directly return `thisCollaboration`, but this change will be more friendly to Scala 2.13 due to  removing `toSeq` can save a memory copy in Scala 2.13. I am checking the correctness and performance benefits now, maybe give another pr later.